File: backend/app/schemas/evaluation.py
import uuid
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field


# --- Pronunciation ---

class PhonemeScoreOut(BaseModel):
    word: str
    word_score: int | None = None
    phoneme: str
    phoneme_score: int
    is_error: bool = False
    suggested_phoneme: str | None = None
    start_time_ms: int | None = None
    end_time_ms: int | None = None

    class Config:
        from_attributes = True


class WordScoreOut(BaseModel):
    word: str
    score: int
    phonemes: list["PhonemeScoreOut"] | None = None

    class Config:
        from_attributes = True


class ProsodyOut(BaseModel):
    intonation_score: int | None = None
    rhythm_score: int | None = None
    stress_errors: list[dict] | None = None


class PronunciationEvaluateRequest(BaseModel):
    reference_text: str = Field(..., min_length=1, max_length=2000)
    language: str = "en-US"
    detail_level: str = "full"  # "basic" or "full"


class PronunciationEvaluateResponse(BaseModel):
    overall_score: int
    pronunciation_score: int | None = None
    fluency_score: int | None = None
    completeness_score: int | None = None
    words: list[WordScoreOut] | None = None
    prosody: ProsodyOut | None = None
    advice: str | None = None


class PronunciationDetailResponse(BaseModel):
    utterance_id: uuid.UUID
    overall_score: int
    pronunciation_score: int | None = None
    fluency_score: int | None = None
    completeness_score: int | None = None
    words: list[WordScoreOut] | None = None
    prosody: ProsodyOut | None = None
    advice: str | None = None
    evaluated_at: datetime | None = None


# --- Grammar ---

class GrammarErrorOut(BaseModel):
    utterance_id: uuid.UUID
    original: str
    error_type: str
    error_span: dict  # {"start": int, "end": int}
    correction: str
    corrected_sentence: str | None = None
    explanation: str | None = None
    severity: str = "medium"

    class Config:
        from_attributes = True


class GrammarReportResponse(BaseModel):
    session_id: uuid.UUID
    errors: list[GrammarErrorOut] = []
    optimization_suggestions: list[dict] = []
    total_errors: int = 0
    total_suggestions: int = 0


class GrammarCorrectRequest(BaseModel):
    text: str = Field(..., min_length=1, max_length=2000)
    mode: str = "full"  # "light" or "full"


class GrammarCorrectResponse(BaseModel):
    original: str
    errors: list[GrammarErrorOut] = []
    corrected_text: str | None = None
    optimization_suggestions: list[dict] = []

File: frontend/src/api/evaluation.ts
import apiClient from './client';

// --- Phoneme Score ---
export interface PhonemeScore {
  word: string;
  word_score: number | null;
  phoneme: string;
  phoneme_score: number;
  is_error: boolean;
  suggested_phoneme: string | null;
  start_time_ms: number | null;
  end_time_ms: number | null;
}

// --- Word Score ---
export interface WordScore {
  word: string;
  score: number;
  phonemes: PhonemeScore[] | null;
}

// --- Prosody ---
export interface Prosody {
  intonation_score: number | null;
  rhythm_score: number | null;
  stress_errors: Record<string, unknown>[] | null;
}

// --- Pronunciation Evaluate (V1.1) ---
export interface PronunciationEvaluateResult {
  overall_score: number;
  pronunciation_score: number | null;
  fluency_score: number | null;
  completeness_score: number | null;
  words: WordScore[] | null;
  prosody: Prosody | null;
  advice: string | null;
  reference_audio_url: string | null;
}

// --- Pronunciation Detail ---
export interface PronunciationDetail {
  utterance_id: string;
  overall_score: number;
  pronunciation_score: number | null;
  fluency_score: number | null;
  completeness_score: number | null;
  words: WordScore[] | null;
  prosody: Prosody | null;
  advice: string | null;
  evaluated_at: string | null;
}

// --- Grammar Correction ---
export interface GrammarErrorOut {
  utterance_id: string;
  original: string;
  error_type: string;
  error_span: Record<string, number>;
  correction: string;
  corrected_sentence: string | null;
  explanation: string | null;
  severity: string;
}

export interface GrammarCorrectResult {
  original: string;
  errors: GrammarErrorOut[];
  corrected_text: string | null;
  optimization_suggestions: Record<string, unknown>[];
}

export interface GrammarReportResult {
  session_id: string;
  errors: GrammarErrorOut[];
  optimization_suggestions: Record<string, unknown>[];
  total_errors: number;
  total_suggestions: number;
}

export const evaluationApi = {
  /** V1.1: Evaluate pronunciation for a single sentence */
  evaluatePronunciation(audio: File, referenceText: string, language = 'en-US', detailLevel = 'full') {
    const formData = new FormData();
    formData.append('audio', audio);
    formData.append('reference_text', referenceText);
    formData.append('language', language);
    formData.append('detail_level', detailLevel);
    return apiClient
      .post<PronunciationEvaluateResult>('/pronunciation/evaluate', formData, {
        headers: { 'Content-Type': 'multipart/form-data' },
        timeout: 30000,
      })
      .then((res) => res.data);
  },

  getPronunciationDetail(sessionId: string, utteranceId: string) {
    return apiClient
      .get<PronunciationDetail>(`/sessions/${sessionId}/pronunciation/${utteranceId}`)
      .then((res) => res.data);
  },

  /** V1.1: Get grammar error report for a session */
  getGrammarReport(sessionId: string) {
    return apiClient
      .get<GrammarReportResult>(`/sessions/${sessionId}/grammar-report`)
      .then((res) => res.data);
  },

  /** V1.1: Instant grammar correction */
  correctGrammar(text: string, mode: 'light' | 'full' = 'full') {
    return apiClient
      .post<GrammarCorrectResult>('/grammar/correct', { text, mode })
      .then((res) => res.data);
  },
};
